File: 142-linked-list-cycle-ii/linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head==None or head.next==None:
            return None
        fast=head
        slow=head
        while(fast!=None and fast.next!=None):
            fast=fast.next.next
            slow=slow.next
            if fast==slow:
                break
        else:
            return None
            
        while head!=fast:
            head=head.next
            fast=fast.next
        return head
        

File: 142-linked-list-cycle-ii/linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if(head==NULL || head->next==NULL)
        return NULL;
        // first check there is a cycle or not 
        // nigga it is imp for edge cases where there is no cycle
        
        ListNode* slow=head;
        ListNode* fast=head;
        bool check=false;
        // Floyd's algo 
        while(fast!=NULL && fast->next!=NULL){
            slow=slow->next;
            fast=fast->next->next;
            if(slow==fast){
                check=true;
                break;
            }
        }
        if(!check)return NULL;
        while(head!=slow){
            head=head->next;
            slow=slow->next;
        }
        return head;
        
    }
};